examples/codes/external_code.ex.py

```diff
@@ -0,0 +1,428 @@
+
+
+@dataclass
+class ECInputs:
+    """External Code Inputs"""
+
+    param1: float = None
+    param2: float = 6
+
+
+@dataclass
+class ECOutputs:
+    """External Code Outputs"""
+
+    param1: float = None
+    param2: float = None
+
+
+# %% [markdown]
+# ## Linking the code to bluemira
+#
+# To link an external code to bluemira you need a few bits of machinery
+#
+# * A `MappedParameterFrame` that links bluemira parameter names and units to the
+#   external code
+# * A `RunMode` class to specify the possibly running modes
+# * A `Solver` that orchistrates the running of the code
+# * Some task for the `Solver` to run typically there are three:
+#     * A `Setup` task which writes the input file for the code
+#     * A `Run` task which runs the code
+#     * A `Teardown` task which reads the output file of the code
+#
+# The `MappedParameterFrame` here gets the defaults and sets the mappings.
+# Notice that "param2" is sent to the code but "param1" is not.
+
+# %%
+@dataclass
+class ECParameterFrame(MappedParameterFrame):
+    """External Code ParameterFrame"""
+
+    header: Parameter[bool]
+    line_number: Parameter[bool]
+    param1: Parameter[float]
+    param2: Parameter[float]
+
+    _defaults = (ECOpts(), ECInputs())
+
+    _mappings = {
+        "header": ParameterMapping("add_header", send=True, recv=False),
+        "line_number": ParameterMapping("number", send=True, recv=False),
+        "param1": ParameterMapping("param1", send=False, recv=True, unit="MW"),
+        "param2": ParameterMapping("param2", send=True, recv=True, unit="GW"),
+    }
+
+    @property
+    def mappings(self) -> Dict:
+        """Code Mappings"""
+        return self._mappings
+
+    @classmethod
+    def from_defaults(cls) -> MappedParameterFrame:
+        """Setup from defaults"""
+        dd = {}
+        for _def in cls._defaults:
+            dd = {**dd, **asdict(_def)}
+        return super().from_defaults(dd)
+
+
+# %%
+class RunMode(BaseRunMode):
+    """
+    RunModes for external code
+    """
+
+    RUN = auto()
+    READ = auto()
+    MOCK = auto()
+
+
+# %% [markdown]
+# The `Setup` class pulls over the inputs from bluemira as described by the
+# mapping and creates the input file. The output of the run method returns
+# the commandline options list.
+
+# %%
+class Setup(CodesSetup):
+    """Setup task"""
+
+    params: ECParameterFrame
+
+    def __init__(self, params: ParameterFrame, problem_settings: Dict, infile: str):
+        super().__init__(params, PRG_NAME)
+        self.problem_settings = problem_settings
+        self.infile = infile
+
+    def update_inputs(self) -> Dict:
+        """Update inputs from bluemira"""
+        self.inputs = ECInputs()
+        self.options = ECOpts()
+        inp = self._get_new_inputs()
+        for thing in (self.problem_settings.items(), inp.items()):
+            for k, v in thing:
+                if k in self.options.__annotations__:
```

Comment:
Is  `[f.name for f in dataclasses.fields(self.options]` a more idiomatic way of retrieving this list?

Also, is my understanding right, in that we're doing two different things with `options` and `inputs`, and that's what this `if` is switching on? If so, I think it may be clearer to just have two separate loops, then we don't need the `if` at all

Reply:
the options and inputs can come from either problem_settings or inp but I've made it a bit clearer